Add --issues option to report issues instead of PRs.

jcvernaleo opened the pull request:
This PR isn't 100% done yet but I wanted to show you to see if you were interested before I went further.  It is useful for me, but you might consider it out of scope.

The main cleanup it needs is changing some references in the code from PR to something a little more generic (item maybe) since those things can now be PRs or issues.  Other than cleanup like that however, it seems to work just fine for me for reporting issues or PRs.

jcvernaleo commented:
Just pinging to see if there were any thoughts on this PR.  I would like to know if you though this was a good way to go before I fix the travis issues and clean things up a bit more.

If you do think it is out of scope I could either spin it off as a new cmd or whole new project since it is useful to me at least.

ryan-roemer commented:
@jcvernaleo -- Apologies, I'm just back from a week's vacation and wading through a ton of my work backlog. Let me think about this and ping you back here hopefully next week (feel free to ping me again if I'm not responding).

In the meantime, let's consider whether we want to view this as an opportunity to display:

* PRs (default)
* Issues
* PRs _and_ Issues

?

jcvernaleo commented:
No worries and no rush.  I'll be at AsiaBSDCon next week so I plan on falling pretty far behind in things myself.

jcvernaleo commented:
So I'm finally caught back up from Japan and ready to do things again.
Any further thoughts as to what would be the best direction to go with this PR?

ryan-roemer commented:
I'm thinking that we add the flags:

* `--issues` default `false`. Add issues to report.
* `--pull-requests` default `true`. Add pull requests to report.
* Error on setup if not 1+ of the above options.

Then refactor the code with something like this pseudo-code (_warning_: completely untested in any fashion)

```js
      // TODO: Add opts.issues, opts.pullRequests to options object for class.
      // TODO: Try not to reference the `program` object here.

      // Iterate repositories
      async.each(results.repos, function (repo, repoCb) {
        // Iterate type of item to request.
        async.parallel([
          // Type: Pull Requests
          function (typeCb) {
            if (!opts.pullRequests) { return typeCb(); }

            github.pullRequests.getAll({
              user: opts.org,
              repo: repo.name,
              state: opts.state,
              per_page: 100 // eslint-disable-line camelcase
            }, function (err, prs) {
              if (prs && prs.length) {
                delete prs.meta;
                repos[repo.name].prs = prs;
              }

              return typeCb(err);
            });
          },

          // Type: Issues
          function (typeCb) {
            if (!opts.issues) { return typeCb(); }

            github.issues.repoIssues({
              user: opts.org,
              repo: repo.name,
              state: opts.state,
              per_page: 100 // eslint-disable-line camelcase
            }, function (err, issues) {
              if (issues issues prs.length) {
                delete issues.meta;
                // TODO: Refactor reporting to handle `issues` like `prs`
                // TODO: (OR) refactor `prs` above and `issues` here to something generic.
                repos[repo.name].issues = issues;
              }

              return typeCb(err);
            });
          }

        ], repoCb); 

      }, function (err) {
        return cb(err, repos);
      });
```

And then do all the other refactoring to:

* The reports
* The docs

Does that seem like a sensible / workable approach?

And with all that expansion of scope does it sound like something you're still up for?

Thanks!

jcvernaleo commented:
I think that sounds like a good way to go.  I probably lack the time today to work on it (unless I get really frustrated with my real work for the day of course) but I'll try to take a shot at it over the weekend or on Monday.

jcvernaleo commented:
I updated the PR to do things more like you suggested.
I just have one issue that I'm not sure how you want to deal with.
If I add an option --pull-reports that defaults to true then as far as I can tell you cannot turn it off without another different option (since commander does seem to automatically give you --no-OPTION as far as I can tell).  So instead I added --nopr to suppress printing PRs for cases where you only want issues and --issues for cases where you want issues added in.

jcvernaleo commented:
Oh and sorry for taking so long to do it.  The coding didn't take very long but finding the time to sit and do it did take a while.

jcvernaleo commented:
@ryan-roemer did you ever get a chance to look at this?

ryan-roemer commented on README.md:
```diff
@@ -37,6 +37,8 @@ $ pull-report --help
     --gh-token <token>    GitHub token
     --pr-url              Add pull request URL to output
     --repo-type <type>    Repo type (default: all|member|private)
+    -pull-requests        Display pull requests
```
`--pull-requests`

ryan-roemer replied:
Add default note.

ryan-roemer commented on README.md:
```diff
@@ -37,6 +37,8 @@ $ pull-report --help
     --pr-url              Add pull request URL to output
     --repo-type <type>    Repo type (default: all|member|private)
+    -pull-requests        Display pull requests
+    -issues               Display issues
```
`--issues`

ryan-roemer replied:
Add default note

ryan-roemer commented on pull-report.js:
```diff
@@ -167,8 +204,10 @@ if (require.main === module) {
-    .option("--pr-url", "Add pull request URL to output", false)
+    .option("--pr-url", "Add pull request or issue URL to output", false)
     .option("--repo-type <type>", "Repo type (default: all|member|private)", "all")
+    .option("--noprs", "Do not display pull requests", false)
```
This doesn't seem to match the README.md documentation. 

I think you can just add `--pull-requests` with default `true` and updated message. And I _think_ the `--no-pull-requests` is then automagically provided, but we should check this.

ryan-roemer commented:
@jcvernaleo -- Great work here! There are a few incongruities with the command line options / docs and a few other nits. Ping me when those are taken care of / other questions, and hopefully next iteration we ship!

jcvernaleo commented:
@ryan-roemer I went through nodejs commander pretty carefully and it does not give you a --no-OPTION for free.  The only magic it provides is if you define an option like --no-pull-requests then you get the variable program.pullRequests=true by default.  So in light of that it doesn't seem to make sense to have an bool option that defaults to true.  So I tried to just make things consistent while having a --no-pull-requests option.  Let me know if that seems good to you.

ryan-roemer commented:
I actually switched it to `--issue-type (pull-request|issue)` given the commander limitations. Thanks for this great work!

ryan-roemer commented:
Released in `pull-report@0.3.0`
